### PR TITLE
[PowerPC] Fix shuffle combine with undef elements

### LINKED ISSUE #77748: Miscompile of dav1d on ppc64le
The checkasm test of https://code.videolan.org/videolan/dav1d fails on ppc64le when built with `meson --build-type=plain`.

The issue bisects to 1a2e77cf9e11dbf56b5720c607313a566eebb16e, which presumably just exposes the issue.

Using opt-bisect-limit and debug-counter=instcombine-visit-count we can narrow this down to an InstCombine transform. However, I believe the transform itself is correct (it just converts some extractelement/insertelement into a shufflevector) and this just exposes an issue in the backend.

The before.ll and after.ll in https://gist.github.com/nikic/b4c015523bcd22ea159d183c0648deea are before and after the transform.

The llc output for these two files is before.s and after.s in https://gist.github.com/nikic/ea9909b8216726762c8f2f277abc8006. (I've run `llc` with `opt-bisect-limit=0` to mostly isolate isel only.)

The difference is in two vperms and corresponding constant pool shuffle indices.

### PR DESCRIPTION
**Pull request by nikic**
This custom DAG combine works on a shuffle where one source vector is a zero splat, which means we can adjust the shuffle indices to refer to any element of the splat -- as long as we stay in the same vector.

In the case where an undef (-1) index into the non-splat vector was used, we ended up adjusting the splat index to -1+NumElements, which points into the wrong vector.

Fix this by not adjusting indices where the other one is undef. There are four cases this theoretically affects, but in practice I only mangaged to demonstrate a miscompile with one of them. I think two of theses are effectively dead due to the operand canonicalization at the start of the transform.

Fixes https://github.com/llvm/llvm-project/issues/77748.

### CODE REVIEW

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 1]
```diff
@@ -15428,15 +15428,15 @@ SDValue PPCTargetLowering::combineVectorShuffle(ShuffleVectorSDNode *SVN,
     // vector_shuffle<0,16,1,17,2,18,3,19,4,20,5,21,6,22,7,23> t1, <zero>
     if (Mask[0] < NumElts)
       for (int i = 1, e = Mask.size(); i < e; i += 2) {
-        if (ShuffV[i] < 0)
+        if (ShuffV[i] < 0 || ShuffV[i - 1] < 0)
```
Comment (bzEq) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:15431]: The change looks good to me(keeping the index as it is if `undef` found in consecutive elements). However, I think what @nemanja suggested in https://github.com/llvm/llvm-project/issues/77748#issuecomment-1887670450 might be potential better for later instruction selection. That's to say, the more same elements in the mask, the more chance we can use single PPC vector splat instruction to materialize the mask immediate.

Comment (nikic) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp]: Done!

**Review by stefanp-synopsys — APPROVED** [review]
I think this looks good!
Thank you for fixing this issue.

**Review by bzEq — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**